### PR TITLE
(GH-225) Add custom_insync to rsapi docs

### PR DESCRIPTION
**Pull request by michaeltlombardi**
This PR coincides with the work for puppetlabs/puppet-resource_api#225
to add the new custom_insync feature and functionality to rsapi resources;

It covers a common use case in the code and explains the purpose and some
implementation concerns.

### CODE REVIEW

File: language/resource-api/README.md [hunk 1]
```diff
@@ -356,6 +356,43 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+  name: 'apt_key',
+  features: [ 'custom_insync' ],
+)
+
+# lib/puppet/provider/apt_key/apt_key.rb
+class Puppet::Provider::AptKey::AptKey
+  def insync?(context, name, attribute_name, is_hash, should_hash)
+    case attribute_name
+    when :some_array_attribute
+      if should_hash[:force]
```
Comment (DavidS) [inline language/resource-api/README.md:375]: please do define the attributes used here in the type, so that it is clear how they are defined (property vs parameter).

Since this is also only called for properties, it's `property_name`, not `attribute_name` in the sig. Also needs to be changed in the test in https://github.com/puppetlabs/puppet-resource_api/pull/285

File: language/resource-api/README.md [hunk 2]
```diff
@@ -356,6 +356,43 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+        context.notice("Checking a subset match array")
+        # This will return true so long as all values in the array for
+        # the should_hash are also in the array of the is_hash.
+        return (is_hash[attribute_name] & should_hash[attribute_name]).sort == should_hash[attribute_name].sort
+      end
+    end
+  end
+```
+
+Specifying the `custom_insync` feature flag ensures that all of the property attributes on the type will call the `insync?` method in the provider, if it exists, enabling custom state comparisons as complex or simple as required. The `insync?` method in the provider has access to the `context` for logging, the name of the resource being checked, the attribute that is currently being inspected for insync status, and hashes representing the current state of the resource as well as the canonicalized should state of the resource.
```
Comment (DavidS) [inline language/resource-api/README.md:390]:
```suggestion
When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties. This can be used to implement custom state comparisons, when the default value-equality is not sufficient. The `insync?` method receives a `context` for logging, the `name` of the resource being checked, the `property_name` that is currently being inspected for insync status, the current state of the resource in `is_hash`, as well as the desired state (canonicalized) in `should_hash`.
```

File: language/resource-api/README.md [hunk 3]
```diff
@@ -356,6 +356,43 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+
+The `insync?` method is only and always triggered for each of the property attributes of the type.
```
Comment (DavidS) [inline language/resource-api/README.md:392]:
```suggestion
The `insync?` method is executed exactly once for each of the properties of the type per resource instance.
The call happens during application of the catalog when puppet decides whether or not it needs to enforce state on a specific resource.
```

File: language/resource-api/README.md [hunk 4]
```diff
@@ -356,6 +356,43 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+
+If the custom `insync?` method in the provider must return a truthy or falsey result, or `nil`; if the return is `nil` the comparison defaults to the existing comparison logic for properties in Puppet. This means that unless the `insync?` method in the provider has special handling for the attribute Puppet will default to the existing state comparison behaviors. Therefore, only properties which need special handling need to be addressed in the `insync?` method in the provider.
```
Comment (DavidS) [inline language/resource-api/README.md:394]:
```suggestion
The `insync?` method's return value will be interpreted as truthy or falsey, following the regular ruby rules, except `nil` values; if the return is `nil` the comparison defaults to the existing comparison logic for properties in Puppet. This means that unless the `insync?` method in the provider has special handling for the attribute Puppet will default to the existing state comparison behaviors. Therefore, only properties which need special handling need to be addressed in the `insync?` method in the provider.
```

**Review by DavidS — CHANGES_REQUESTED** [review]

File: language/resource-api/README.md [hunk 5]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+    end
+end
+```
+
+When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
+This can be used to implement custom state comparisons, when the default value-equality is not sufficient.
+The `insync?` method receives a `context` for logging, the `name` of the resource being checked, the `property_name` that is currently being inspected for insync status, the current state of the resource in `is_hash`, as well as the desired state (canonicalized) in `should_hash`.
+
+The `insync?` method is executed exactly once for each of the properties of the type per resource instance.
+The call happens during application of the catalog when puppet decides whether or not it needs to enforce state on a specific resource.
```
Comment (DavidS) [inline language/resource-api/README.md:419]:
```suggestion
The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.
```

File: language/resource-api/README.md [hunk 6]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+
+The `insync?` method's return value will drive the behavior of Puppet for that property:
+
+- If `insync?` returns `nil`, property comparison is handed back to Puppet's built-in property insync checking.
```
Comment (DavidS) [inline language/resource-api/README.md:423]:
```suggestion
- If `insync?` returns `nil`, property comparison is handed back to the runtime environment's built-in property value equality checking.
```

File: language/resource-api/README.md [hunk 7]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+  - This means that _only_ properties which actually need custom insync logic need to be handled in the `insync?` method on the provider. Everything else can be ignored, which will return `nil` and use default comparison logic.
```
Comment (DavidS) [inline language/resource-api/README.md:424]:
```suggestion
  - This means that the provider _only_ needs to deal with properties which actually need custom insync logic. All other properties can be ignored, which will return ruby's default `nil` return value and use default comparison logic.
```

File: language/resource-api/README.md [hunk 8]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+- If `insync?` returns `true`, the property is treated as insync.
+- If `insync?` returns `false`, the property is treated as out of sync and will surface the default Puppet change reporting.
```
Comment (DavidS) [inline language/resource-api/README.md:426]:
```suggestion
- If `insync?` returns `false`, the property is treated as out of sync and will cause the default change reporting.
```

File: language/resource-api/README.md [hunk 9]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+  - e.g. `changed <is value> to <should value>`
+- If `insync?` returns a string, the property is treated as out of sync and the string is surfaced in the Puppet change reporting.
```
Comment (DavidS) [inline language/resource-api/README.md:428]:
```suggestion
- If `insync?` returns a string, the property is treated as out of sync and the string is used for change reporting instead of the default. This can be used to provide additional information about the reason for a change.
```

File: language/resource-api/README.md [hunk 10]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+The `insync?` method is executed exactly once for each of the properties of the type per resource instance.
+The call happens during application of the catalog when puppet decides whether or not it needs to enforce state on a specific resource.
+
+The `insync?` method's return value will drive the behavior of Puppet for that property:
+
+- If `insync?` returns `nil`, property comparison is handed back to Puppet's built-in property insync checking.
+  - This means that _only_ properties which actually need custom insync logic need to be handled in the `insync?` method on the provider. Everything else can be ignored, which will return `nil` and use default comparison logic.
+- If `insync?` returns `true`, the property is treated as insync.
+- If `insync?` returns `false`, the property is treated as out of sync and will surface the default Puppet change reporting.
+  - e.g. `changed <is value> to <should value>`
```
Comment (DavidS) [inline language/resource-api/README.md:427]:
```suggestion
  - e.g. `Notice: /Stage[main]/Main/File[/tmp/foo]/mode: mode changed <is value> to <should value>`
```

File: language/resource-api/README.md [hunk 11]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+- If `insync?` returns a string, the property is treated as out of sync and the string is surfaced in the Puppet change reporting.
+  - In the `test_custom_insync` provider example above, you would get the message `Adding missing members ['SomeValue']` if `SomeValue` was specified as a member of `some_array` but not found in the actual state returned from `get`.
```
Comment (DavidS) [inline language/resource-api/README.md:429]:
```suggestion
  - In the `test_custom_insync` provider example above, you would get the message `Notice: /Stage[main]/Main/Test_custom_insync[something]/some_array: Adding missing members ['SomeValue']` if `SomeValue` was specified as a member of `some_array` but not found in the actual state returned from `get`.
```

File: language/resource-api/README.md [hunk 12]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+    changes.each do |_name, is_and_should|
+      context.notice("Setting with #{is_and_should[:should].inspect}")
+    end
+  end
+
+  def insync?(context, name, _property_name, is_hash, should_hash)
+    context.notice("Checking whether #{name} is out of sync")
+
+    return true unless should_hash[:force]
+    context.notice("Out of sync!")
```
Comment (DavidS) [inline language/resource-api/README.md:483]:
```suggestion
    context.debug("Checking whether #{name} is out of sync")

    return true unless should_hash[:force]
    context.debug("Out of sync!")
```

File: language/resource-api/README.md [hunk 13]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
+        return missing_members.empty? ? true : "Adding missing members #{missing_members}"
+      end
+    end
+end
+```
+
+When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
+This can be used to implement custom state comparisons, when the default value-equality is not sufficient.
+The `insync?` method receives a `context` for logging, the `name` of the resource being checked, the `property_name` that is currently being inspected for insync status, the current state of the resource in `is_hash`, as well as the desired state (canonicalized) in `should_hash`.
```
[the remark this answers is not in the document]
Comment (michaeltlombardi) [reply language/resource-api/README.md]: Moved to after the return value explanation.

File: language/resource-api/README.md [hunk 14]
```diff
@@ -356,6 +356,136 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+
+The `insync?` method is executed exactly once for each of the properties of the type per resource instance.
```
[the remark this answers is not in the document]
Comment (michaeltlombardi) [reply language/resource-api/README.md]: Modified to read:

> The `insync?` method is executed exactly once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
> If no properties were specified in the instance of a resource, `insync?` is not called.

[the remark this answers is not in the document]
Comment (michaeltlombardi) [reply language/resource-api/README.md]: wasn't sure on the wording for read-only/init.

File: language/resource-api/README.md [hunk 15]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+      end
+    end
+end
+```
+
+In normal operations, the runtime environment uses strict value equality to determine if a resource instance needs to be state-enforced.
+When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
+This can be used to implement custom state comparisons, when the default value-equality is not sufficient.
+
+The `insync?` method is executed exactly once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
```
Comment (clairecadman) [inline language/resource-api/README.md:425]:
```suggestion
The `insync?` method is executed once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
```

File: language/resource-api/README.md [hunk 16]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+
+The `insync?(context, name, property_name, is_hash, should_hash)` method receives the following arguments:
+
+- a `context` to access the runtime environment,
+- the `name` of the resource being checked,
+- the `property_name` that is currently being inspected for insync status,
+- the current state of the resource in `is_hash`, and
+- the desired state (canonicalized) in `should_hash`.
+
+For resources which specify the `custom_insync` feature flag but have _no_ defined properties the Resource API adds a spectial hidden property, `rsapi_custom_insync_trigger`, which ensures that `insync?` is still called for the resource.
```
Comment (clairecadman) [inline language/resource-api/README.md:448]:
```suggestion
For resources which specify the `custom_insync` feature flag but have _no_ defined properties, the Resource API adds a spectial hidden property —`rsapi_custom_insync_trigger` —  which ensures that `insync?` is still called for the resource.
```

File: language/resource-api/README.md [hunk 17]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
 > values changed
 ```
 
+### Provider feature: `custom_insync`
+
+Allows for overriding default property `insync?` checking using custom logic in the provider.
```
Comment (clairecadman) [inline language/resource-api/README.md:368]:
```suggestion
The `custom_insyc` feature lets you for override default property `insync?` checking using custom logic in the provider.
```

File: language/resource-api/README.md [hunk 18]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+        context.debug("Checking a subset match array")
+        found_members = (is_hash[property_name] & should_hash[property_name]).sort
+        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
+        return missing_members.empty? ? true : "Adding missing members #{missing_members}"
+      end
+    end
+end
+```
+
+In normal operations, the runtime environment uses strict value equality to determine if a resource instance needs to be state-enforced.
```
Comment (clairecadman) [inline language/resource-api/README.md:421]: Does this mean without using this do feature?

File: language/resource-api/README.md [hunk 19]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
```
Comment (clairecadman) [inline language/resource-api/README.md:422]:
```suggestion
When you specify the `custom_insync` feature flag, `insync?` is called once for each of the properties.
```

File: language/resource-api/README.md (hunk 18 above)
Comment (michaeltlombardi) [reply language/resource-api/README.md:421]: It does.

File: language/resource-api/README.md [hunk 20]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+This can be used to implement custom state comparisons, when the default value-equality is not sufficient.
+
+The `insync?` method is executed exactly once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
+If no properties were specified in the instance of a resource, `insync?` is not called.
+The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.
+
+The `insync?` method's return value will drive the behavior of Puppet for that property:
```
Comment (clairecadman) [inline language/resource-api/README.md:429]:
```suggestion
The `insync?` method's return value changes the way Puppet handles the property. For example:
```

File: language/resource-api/README.md [hunk 21]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+- the current state of the resource in `is_hash`, and
+- the desired state (canonicalized) in `should_hash`.
+
+For resources which specify the `custom_insync` feature flag but have _no_ defined properties the Resource API adds a spectial hidden property, `rsapi_custom_insync_trigger`, which ensures that `insync?` is still called for the resource.
+This hidden property is not used or seen anywhere else in the provider, only in the `insync?` method.
+
+As with all calls to `insync?`, a custom change reporting message can be surfaced by returning a string instead of `true` or `false`.
+If no custom change reporting message is returned as a string, the change reporting in Puppet (if `false` is returned) will always be:
+
+> Custom insync logic determined that this resource is out of sync
```
Comment (clairecadman) [inline language/resource-api/README.md:454]: Is this the message the user would see?

File: language/resource-api/README.md [hunk 22]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+      type:      'Boolean',
+      desc:      'If true, the resource will be treated as not being insync.',
+      behaviour: :parameter,
+      default:   false,
+    },
+  },
+)
+
+# lib/puppet/provider/test_custom_insync_hidden_property/test_custom_insync_hidden_property.rb
+class Puppet::Provider::TestCustomInsyncHiddenProperty::TestCustomInsyncHiddenProperty
```
Comment (clairecadman) [inline language/resource-api/README.md:480]: What is this example showing compared to the first one?

File: language/resource-api/README.md [hunk 23]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+        missing_members = should_hash[property_name].reject { |member| found_members.include?(member) }
+        return missing_members.empty? ? true : "Adding missing members #{missing_members}"
+      end
+    end
+end
+```
+
+In normal operations, the runtime environment uses strict value equality to determine if a resource instance needs to be state-enforced.
+When the `custom_insync` feature flag is specified, `insync?` is called once for each of the properties.
+This can be used to implement custom state comparisons, when the default value-equality is not sufficient.
```
Comment (clairecadman) [inline language/resource-api/README.md:423]:
```suggestion
You can use `custom_insync` to implement custom state comparisons when the default value-equality is not sufficient.
```

File: language/resource-api/README.md [hunk 24]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+
+The `insync?` method is executed exactly once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
+If no properties were specified in the instance of a resource, `insync?` is not called.
+The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.
```
Comment (clairecadman) [inline language/resource-api/README.md:427]: Is this desired state?

File: language/resource-api/README.md (hunk 21 above)
Comment (michaeltlombardi) [reply language/resource-api/README.md:454]: Yes.

File: language/resource-api/README.md (hunk 22 above)
Comment (michaeltlombardi) [reply language/resource-api/README.md:480]: This shows an example _without_ an insyncable property - which adds a hidden property to ensure that `insync?` is triggered.

File: language/resource-api/README.md (hunk 24 above)
Comment (michaeltlombardi) [reply language/resource-api/README.md:427]: It determines if the resource is _in_ the desired state.

File: language/resource-api/README.md (hunk 18 above)
Comment (clairecadman) [reply language/resource-api/README.md:421]: Maybe say "When you don't use "custom_insync, the runtime environment etc...."?

File: language/resource-api/README.md (hunk 22 above)
Comment (clairecadman) [reply language/resource-api/README.md:480]: Ah okay, maybe make that clearer?

File: language/resource-api/README.md [hunk 25]
```diff
@@ -355,6 +363,144 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+You can use `custom_insync` to implement custom state comparisons when the default value-equality is not sufficient.
+
+The `insync?` method is executed once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
+If no properties were specified in the instance of a resource, `insync?` is not called.
+The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.
+
+The `insync?` method's return value changes the way Puppet handles the property. For example:
+
+- If `insync?` returns `nil`, property comparison is handed back to the runtime environment's built-in property value equality checking.
+  - This means that the provider _only_ needs to deal with properties which actually need custom insync logic. All other properties can be ignored, which will return ruby's default `nil` return value and use default comparison logic.
```
Comment (clairecadman) [inline language/resource-api/README.md:432]: Capital R for Ruby.

File: language/resource-api/README.md (hunk 18 above)
Comment (michaeltlombardi) [reply language/resource-api/README.md]: I believe I've addressed this by breaking the docs up a little to clarify:

https://github.com/puppetlabs/puppet-specifications/pull/153/files#diff-23f362101839bf8ed43d75a53178cf3cb35aa2d8169fc0163f3a7619f87ef79fR421-R424

File: language/resource-api/README.md (hunk 22 above)
Comment (michaeltlombardi) [reply language/resource-api/README.md:491]: I've moved this entire section into a subheading, does that address this concern?

https://github.com/puppetlabs/puppet-specifications/pull/153/files#diff-23f362101839bf8ed43d75a53178cf3cb35aa2d8169fc0163f3a7619f87ef79fR457

**Comment by michaeltlombardi** [comment]
@clairecadman just pinging to see if my updates addressed your review comments :)

File: language/resource-api/README.md (hunk 24 above)
Comment (clairecadman) [reply language/resource-api/README.md:428]: Maybe rephrase it like this?

File: language/resource-api/README.md (hunk 22 above)
Comment (clairecadman) [reply language/resource-api/README.md:491]: Yes that is better, thanks!

**Comment by clairecadman** [comment]
@michaeltlombardi yes all good thanks! I don't want to change anything else as it would make it inconsistent with the rest of the README.

File: language/resource-api/README.md [hunk 26]
```diff
@@ -355,6 +363,155 @@ For example, in the Puppet runtime environment this is bound to the `strict` set
...
+When you do not specify the `custom_insync` feature flag, the runtime environment uses strict value equality to determine if a resource instance needs to be state-enforced.
+
+When you **do** specify the `custom_insync` feature flag, `insync?` is called once for each of the properties.
+You can use `custom_insync` to implement custom state comparisons when the default value-equality is not sufficient.
+
+The `insync?` method is executed once for each of the **properties** (not parameters, read-only, or init attributes) of the type per resource instance.
+If no properties were specified in the instance of a resource, `insync?` is not called.
+The call happens during application of the catalog when the runtime environment decides whether or not it needs to enforce state on a specific resource.
+
+The `insync?` method **must** return `nil`, `true` or `false` as the result of the property comparison:
```
Comment (DavidS) [inline language/resource-api/README.md:430]: this is not explaining the new API requested by @joshcooper

Comment (michaeltlombardi) [reply language/resource-api/README.md:430]: I think this is resolved by the immediately following block which explains that the message only matters if the result is actually `false`:

image

**Comment by michaeltlombardi** [comment]
@clairecadman, @DavidS, can we get a final approval and merge on this if there's nothing else?

**Review by DavidS — APPROVED** [review]

**Review by david22swan — APPROVED** [review]
